tests/integration/cattletest/core/test_cluster.py

```diff
@@ -0,0 +1,75 @@
+__author__ = 'sonchang'
+
+from common_fixtures import *  # NOQA
+
+
+def test_create_cluster(admin_client):
+    cluster = admin_client.create_cluster(
+        name='testcluster1', port=9000, discoverySpec='file:///tmp/my_cluster')
+
+    cluster = wait_until_expected_state(admin_client, cluster, 'active')
+
+    assert_fields(cluster, {
+        "kind": "cluster",
+        "port": 9000,
+        "discoverySpec": "file:///tmp/my_cluster",
+        "state": "active"
+    })
+
+
+def test_create_cluster_with_default_discovery_spec(admin_client):
+    cluster = admin_client.create_cluster(
+        name='testcluster2', port=9001)
+
+    cluster = wait_until_expected_state(admin_client, cluster, 'active')
+
+    assert_fields(cluster, {
+        "port": 9001,
+        "state": "active"
+    })
+    # Uncomment after mock external handler is put in place
+    # assert cluster.discoverySpec is not None
+    # assert str(cluster.discoverySpec).startswith('token://')
+
+
+def test_cluster_add_remove_host_actions(admin_client, super_client):
+    cluster = admin_client.create_cluster(
+        name='testcluster3', port=9000)
+
+    cluster = wait_until_expected_state(admin_client, cluster, 'active')
+
+    host1 = super_client.create_host()
+    host1 = super_client.wait_success(host1)
+
+    cluster = cluster.addhost(host=str(host1.id))
+
+    assert len(cluster.hosts()) == 1
+    assert cluster.hosts()[0].id == host1.id
+    assert len(host1.clusters()) == 1
+    assert host1.clusters()[0].id == cluster.id
+
+    try:
+        cluster.addhost(host=str(host1.id))
+        assert False
+    except cattle.ApiError as e:
+        assert e.error.code == 'InvalidReference'
+
+    cluster = cluster.removehost(host=str(host1.id))
+
+    assert len(cluster.hosts()) == 0
+
+    try:
+        cluster = cluster.removehost(host=str(host1.id))
+        assert False
+    except cattle.ApiError as e:
+        assert e.error.code == 'InvalidReference'
+
+    cluster = cluster.addhost(host=str(host1.id))
+    assert len(cluster.hosts()) == 1
+
+    host2 = super_client.create_host()
+    host2 = super_client.wait_success(host2)
+
+    cluster = cluster.addhost(host=str(host2.id))
+
+    assert len(cluster.hosts()) == 2
```

Comment:
Actually, I'll add another test to test the removal of the clusterHostMapping when a host is purged